p2p/shim.go

```diff
@@ -0,0 +1,327 @@
+				msg.Reset()
+
+				wrapper, ok := msg.(Wrapper)
+				if ok {
+					if err := wrapper.Wrap(e.Message); err != nil {
+						rs.Logger.Error(
+							"failed to proxy envelope; failed to wrap message",
+							"ch_id", cs.Descriptor.ID,
+							"msg", e.Message,
+							"err", err,
+						)
+						continue
+					}
+				} else {
+					msg = e.Message
+				}
+
+				bz, err := proto.Marshal(msg)
+				if err != nil {
+					rs.Logger.Error(
+						"failed to proxy envelope; failed to encode message",
+						"ch_id", cs.Descriptor.ID,
+						"msg", e.Message,
+						"err", err,
+					)
+					continue
+				}
+
+				switch {
+				case e.Broadcast:
+					rs.Switch.Broadcast(cs.Descriptor.ID, bz)
+
+				case !e.To.Empty():
+					src := rs.Switch.peers.Get(ID(e.To.String()))
+					if src == nil {
+						rs.Logger.Error(
+							"failed to proxy envelope; failed to find peer",
+							"ch_id", cs.Descriptor.ID,
+							"msg", e.Message,
+							"peer", e.To.String(),
+						)
+						continue
+					}
+
+					if !src.Send(cs.Descriptor.ID, bz) {
+						rs.Logger.Error(
+							"failed to proxy message to peer",
+							"ch_id", cs.Descriptor.ID,
+							"msg", e.Message,
+							"peer", e.To.String(),
+						)
+					}
+
+				default:
+					rs.Logger.Error("failed to proxy envelope; missing peer ID", "ch_id", cs.Descriptor.ID, "msg", e.Message)
+				}
+			}
+		}(cs)
+	}
+}
+
+// handlePeerErrors iterates over each p2p Channel and starts a separate go-routine
+// where we listen for peer errors. For each peer error, we find the peer from
+// the legacy p2p Switch and execute a StopPeerForError call with the corresponding
+// peer error.
+func (rs *ReactorShim) handlePeerErrors() {
+	for _, cs := range rs.Channels {
+		go func(cs *ChannelShim) {
+			for pErr := range cs.PeerErrCh {
+				if !pErr.PeerID.Empty() {
+					peer := rs.Switch.peers.Get(ID(pErr.PeerID.String()))
+					if peer == nil {
+						rs.Logger.Error("failed to handle peer error; failed to find peer", "peer", pErr.PeerID.String())
+						continue
+					}
+
+					rs.Switch.StopPeerForError(peer, pErr.Err)
+				}
+			}
+		}(cs)
+	}
+}
+
+// OnStart executes the reactor shim's OnStart hook where we start all the
+// necessary go-routines in order to proxy peer envelopes and errors per p2p
+// Channel.
+func (rs *ReactorShim) OnStart() error {
+	if rs.Switch == nil {
+		return errors.New("proxyPeerEnvelopes: reactor shim switch is nil")
+	}
+
+	// start envelope proxying and peer error handling in separate go routines
+	rs.proxyPeerEnvelopes()
+	rs.handlePeerErrors()
+
+	return nil
+}
+
+// OnStop executes the reactor shim's OnStop hook where the PeerUpdateCh is
+// closed. The caller must be sure to also stop the real reactor so the shim's
+// proxyPeerEnvelopes can exit successfully. Note, the inbound InCh is explicitly
+// closed on Receive once the real reactor stops and closes all of its p2p
+// Channels.
+func (rs *ReactorShim) OnStop() {
+	close(rs.PeerUpdateCh)
```

Comment:
This is not safe, as there are separate goroutines that write to the channel.

Reply:
Ahh yes, only `AddPeer` and `RemovePeer` send on `PeerUpdateCh`. These two methods should theoretically not be called after the reactor (shim) is stopped.

However, it is also theoretically possible if there is nothing listening on the other end of `PeerUpdateCh` or if that listener is blocked or taking a long time, then those shim `AddPeer` and `RemovePeer` goroutines would panic as you've pointed out.

That being said, I do not immediately see a clean way of handling this. We can either not close this channel or use locks maybe to check before sending?

Reply:
Yeah, so in ADR-061 the router has a `PeerUpdates(context.Context)` method to start a peer update subscription, which it must keep consuming until the context is cancelled. There are other schemes we can use for this as well, this was the first that came to mind.

There needs to be a separate peer updates channel per consumer (because Go does not have built-in support for broadcast channels, so the router has to iterate across all of the subscriptions and duplicate the update event), and we need to know when these should no longer be updated. There should probably be a way to opt-in to this too, since not every reactor may need updates.

Reply:
> There needs to be a separate peer updates channel per consumer 

Currently, this is the case. The shim is the sole producer, the real reactor is the sole consumer.

Reply:
Removed. We no longer close this channel at all.